src/aead/aes/ffi.rs

```diff
@@ -0,0 +1,211 @@
+// Copyright 2018-2024 Brian Smith.
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY
+// SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+// OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+// CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+use super::{Block, KeyBytes, BLOCK_LEN};
+use crate::{bits::BitLength, c, cpu, error, polyfill::slice};
+use core::{num::NonZeroUsize, ops::RangeFrom};
+
+/// nonce || big-endian counter.
+#[repr(transparent)]
+pub(in super::super) struct Counter(pub(super) [u8; BLOCK_LEN]);
+
+// Keep this in sync with AES_KEY in aes.h.
+#[repr(C)]
+#[derive(Clone)]
+pub(in super::super) struct AES_KEY {
+    pub rd_key: [u32; 4 * (MAX_ROUNDS + 1)],
+    pub rounds: c::uint,
+}
+
+// Keep this in sync with `AES_MAXNR` in aes.h.
+const MAX_ROUNDS: usize = 14;
+
+impl AES_KEY {
+    #[inline]
+    pub(super) unsafe fn new(
+        f: unsafe extern "C" fn(*const u8, BitLength<c::int>, *mut AES_KEY) -> c::int,
+        bytes: KeyBytes<'_>,
+        _cpu_features: cpu::Features,
+    ) -> Result<Self, error::Unspecified> {
+        let mut key = Self {
+            rd_key: [0; 4 * (MAX_ROUNDS + 1)],
+            rounds: 0,
+        };
+
+        let (bytes, key_bits) = match bytes {
+            KeyBytes::AES_128(bytes) => (&bytes[..], BitLength::from_bits(128)),
+            KeyBytes::AES_256(bytes) => (&bytes[..], BitLength::from_bits(256)),
+        };
+
+        // Unusually, in this case zero means success and non-zero means failure.
+        if 0 == unsafe { f(bytes.as_ptr(), key_bits, &mut key) } {
+            debug_assert_ne!(key.rounds, 0); // Sanity check initialization.
+            Ok(key)
+        } else {
+            Err(error::Unspecified)
```

Comment:
It is expected that this isn't reached because we've met all the preconditions to ensure the function won't fail.